### PR TITLE
Pay block reward

### PR DESCRIPTION
**Pull request by girazoki**
This PR expands on https://github.com/witnet/witnet-ethereum-block-relay/pull/29 to be able to pay the block relayer. The block reward is divided in two, one for the inclusion and one for the reward.

 Missing things:

- Upgrade Request should also include an upgrade for the block relay reward.
- More extensive testing

### CODE REVIEW

File: contracts/UsingWitnet.sol [hunk 1]
```diff
@@ -33,9 +33,10 @@ contract UsingWitnet {
   }
 
   // Ensures that user-specified rewards are equal to the total transaction value to prevent users from burning any excess value
-  modifier validRewards(uint256 _requestReward, uint256 _resultReward) {
-    require(_requestReward + _resultReward >= _requestReward, "The sum of rewards overflows");
-    require(msg.value == _requestReward + _resultReward, "Transaction value should equal the sum of rewards");
+  modifier validRewards(uint256 _requestReward, uint256 _resultReward, uint256 _blockReward) {
+    require(_requestReward + _resultReward + _blockReward >= _requestReward, "The sum of rewards overflows");
```
Comment (tmpolaczyk) [inline contracts/UsingWitnet.sol:37]: This overflow check is wrong, with the values `_requestReward = 1`, `_resultReward = _blockReward = 2**256 - 1` it returns true:

    require(1 + (2**256 - 1) + (2**256 - 1) >= 1)

So use a different overflow check, and add tests.

**Comment by clbartoli** [comment]
If I got it right, the `blockReward` is defined when posting the data request as the sum of the `inclusionReward` and the `tallyReward`. If this is correct, then there is no need to include the `blockReward` as input in any function, since it can be calculated from the other rewards.

Having this in mind, some modifiers change
- `payingEnough` in WitnetRequestBoard.sol should make sure the value is greater or equal `2*_tally`, the reward of the tally and  the block reward.
- `ValidRewards` in UsingWitnet.sol has to consider double the rewards, something like 
  `require(_requestReward + _resultReward >= _requestReward && 2*(_requestReward +_requestReward) >= 2*_requestReward, "The sum of rewards overflows");

The tests will need to be changed as well, as some of the values of the transactions are not longer valid.`

File: contracts/UsingWitnet.sol [hunk 2]
```diff
@@ -47,12 +48,12 @@ contract UsingWitnet {
   * @param _resultReward Reward specified for the user which posts back the request result
   * @return Sequencial identifier for the request included in the WitnetRequestsBoard
   */
-  function witnetPostRequest(Request _request, uint256 _requestReward, uint256 _resultReward)
+  function witnetPostRequest(Request _request, uint256 _requestReward, uint256 _resultReward, uint256 _blockReward)
```
Comment (mariocao) [inline contracts/UsingWitnet.sol:51]: The `@param` documentation is missing :)

File: contracts/WitnetRequestsBoard.sol [hunk 3]
```diff
@@ -166,10 +167,10 @@ contract WitnetRequestsBoard is WitnetRequestsBoardInterface {
   /// @param _requestAddress The request contract address which includes the request bytecode.
   /// @param _tallyReward The amount of value that will be detracted from the transaction value and reserved for rewarding the reporting of the final result (aka tally) of the data request.
   /// @return The unique identifier of the data request.
-  function postDataRequest(address _requestAddress, uint256 _tallyReward)
+  function postDataRequest(address _requestAddress, uint256 _inclusionReward, uint256 _tallyReward)
```
Comment (mariocao) [inline contracts/WitnetRequestsBoard.sol:170]: Missing `@param` documentation

File: contracts/WitnetRequestsBoard.sol [hunk 4]
```diff
@@ -198,7 +200,7 @@ contract WitnetRequestsBoard is WitnetRequestsBoardInterface {
   /// @dev Increments the rewards of a data request by adding more value to it. The new request reward will be increased by msg.value minus the difference between the former tally reward and the new tally reward.
   /// @param _id The unique identifier of the data request.
   /// @param _tallyReward The new tally reward. Needs to be equal or greater than the former tally reward.
-  function upgradeDataRequest(uint256 _id, uint256 _tallyReward)
+  function upgradeDataRequest(uint256 _id,  uint256 _tallyReward)
```
Comment (mariocao) [inline contracts/WitnetRequestsBoard.sol:203]:
```suggestion
  function upgradeDataRequest(uint256 _id, uint256 _tallyReward)
```

File: contracts/WitnetRequestsBoard.sol [hunk 5]
```diff
@@ -274,6 +276,10 @@ contract WitnetRequestsBoard is WitnetRequestsBoardInterface {
       _epoch,
       _index,
       requests[_id].drOutputHash), "Invalid PoI");
+    address payable relayer = payable(blockRelay.readRelayerAddress(_blockHash, _epoch));
+    uint256 relayerPayment = requests[_id].blockReward/2;
```
Comment (mariocao) [inline contracts/WitnetRequestsBoard.sol:280]: Are relay DR inclusion and block similar in terms of gas cost?
Just asking myself if the factor 2 makes sense or could be improved...

Comment (mariocao) [inline contracts/WitnetRequestsBoard.sol:280]:
```suggestion
    uint256 relayerPayment = requests[_id].blockReward / 2;
```

File: contracts/WitnetRequestsBoard.sol [hunk 6]
```diff
@@ -274,6 +276,10 @@ contract WitnetRequestsBoard is WitnetRequestsBoardInterface {
...
+    requests[_id].blockReward = requests[_id].blockReward - relayerPayment; 
+    relayer.transfer(relayerPayment);
```
Comment (mariocao) [inline contracts/WitnetRequestsBoard.sol:282]:
```suggestion
    address payable relayer = payable(blockRelay.readRelayerAddress(_blockHash, _epoch));
    uint256 relayerPayment = requests[_id].blockReward/2;
    requests[_id].blockReward = requests[_id].blockReward - relayerPayment; 
    relayer.transfer(relayerPayment);
```
Add comment explaining that we are storing reward and doing transfer :)

File: contracts/mocks/MockWitnetRequestsBoard.sol [hunk 7]
```diff
@@ -39,7 +40,7 @@ contract MockWitnetRequestsBoard is WitnetRequestsBoardInterface {
   /// @param _requestAddress The request contract address which includes the request bytecode.
   /// @param _tallyReward The amount of value that will be detracted from the transaction value and reserved for rewarding the reporting of the final result (aka tally) of the data request.
   /// @return The unique identifier of the data request.
-  function postDataRequest(address _requestAddress, uint256 _tallyReward)
+  function postDataRequest(address _requestAddress, uint256 _inclusionReward, uint256 _tallyReward)
```
Comment (mariocao) [inline contracts/mocks/MockWitnetRequestsBoard.sol:43]: Missing `@param` doc

File: contracts/WitnetRequestsBoard.sol (hunk 5 above)
Comment (girazoki) [reply contracts/WitnetRequestsBoard.sol:285]: In theory they are both block inclusions (one for DR and the other one for result), so they should be yes.

**Review by mariocao — APPROVED** [review]

**Pull request merged**